internal/tunnel/argo.go

```diff
@@ -110,32 +116,89 @@ func (mgr *ArgoTunnelManager) Active() bool {
 
 func (mgr *ArgoTunnelManager) Start(serviceURL string) error {
 
+	mgr.tunnelConfig.Logger.WithFields(log.Fields{
+		"service":  mgr.Config().ServiceName,
+		"hostname": mgr.Config().ExternalHostname,
+	}).Infof("starting tunnel")
+
 	if serviceURL == "" {
 		return fmt.Errorf("Cannot start tunnel for %s with empty url", mgr.Config().ServiceName)
 	}
 	mgr.tunnelConfig.OriginUrl = serviceURL
 
 	placeHolderOnlyConnectedSignal := make(chan struct{})
 	mgr.stopCh = make(chan struct{})
+	mgr.errCh = make(chan error)
 
+	// error handler
+	go func() {
+		mgr.tunnelConfig.Logger.WithFields(log.Fields{
+			"service":  mgr.Config().ServiceName,
+			"hostname": mgr.Config().ExternalHostname,
+		}).Infof("Starting error receiver")
+		for {
+			select {
+			case err, open := <-mgr.errCh:
+				if !open {
+					return
+				}
+				mgr.tunnelConfig.Logger.WithFields(log.Fields{
+					"service":  mgr.Config().ServiceName,
+					"hostname": mgr.Config().ExternalHostname,
+				}).Errorf("Received error (%s) %v", reflect.TypeOf(err), err)
+
+				// the errors are all un-exported from origin package :(
+				// hard to figure out when to retry then
+				switch err.(type) {
+
+				case argoTunnelStopError:
+					mgr.tunnelConfig.Logger.WithFields(log.Fields{
+						"service":  mgr.Config().ServiceName,
+						"hostname": mgr.Config().ExternalHostname,
+					}).Errorf("Stopping tunnel")
+					return
+
+				default:
+					mgr.tunnelConfig.Logger.WithFields(log.Fields{
+						"service":  mgr.Config().ServiceName,
+						"hostname": mgr.Config().ExternalHostname,
+					}).Errorf("Retrying tunnel")
+					mgr.Start(serviceURL)
```

Comment:
you'll want to pair this change with PR #72, calling start from another do routine opens an additional source of race. In fact, you may need to add a restart call to close all channels (a combination of stop and start without releasing the mutex in-between).
 - https://github.com/cloudflare/cloudflare-ingress-controller/pull/72